packages/yarnpkg-core/sources/Controllers.ts

```diff
@@ -0,0 +1,105 @@
+import {Fetcher}                                         from './Fetcher';
+import {MultiFetcher}                                    from './MultiFetcher';
+import {MultiResolver}                                   from './MultiResolver';
+import {Resolver, ResolveOptions, MinimalResolveOptions} from './Resolver';
+import {Descriptor, Locator, DescriptorHash, Package}    from './types';
+
+// The CoreResolver and CoreFetcher classes are controllers for the actual resolvers/fetchers.
+// This description uses a resolver for example, but the logic is the same for fetchers.
+// All resolvers with type 'module' are stored in an array `resolvers.module`
+// All resolvers with anything else (ie. type: 'protocol') are loaded into a MultiResolver
+// and stored in `resolvers.protocol`.
+// The logic is that the first loaded module is consulted first with the following options
+// {next: nextResolver, proto: resolvers.proto}. The resolver can choose to return a response,
+// or consult the next resolver in the chain, or consult the protocol resolvers.
+// `next` is a wrapper function to call the next resolver passing it the resolver after that
+
+export class ResolverController implements Resolver {
+  public readonly type = 'module';
+  private readonly index = new Map<Resolver, Resolver | undefined>(); // map of each module resolver to the next in the chain
+  private readonly protocols = new MultiResolver([]); // a MultiResolver aggregating all protocol resolvers
+  private readonly modules = new Array<Resolver>(); // an array of all modules (mostly to keep track of which to call first)
+
+  constructor(resolvers: Array<Resolver>) {
+    resolvers.forEach( (resolver) => { // categorize each plugin
+      if (resolver && resolver.type === 'module') {
+        this.modules.push(resolver);
+      } else if (resolver) {
+        this.protocols.addResolver(resolver);
+      }
+    } );
+    for (let i=0; i < this.modules.length; i++) { // create a map of each module with its next (ie. the chain of module resolvers)
+      this.index.set(this.modules[i], this.modules[i+1]);
+    }
+  }
+
+  prepareNext(func: string, resolver: Resolver | undefined, ...args: any): any {
+    if (resolver) {
+      const next = this.index.get(resolver); // if next is undefined the `prepareNext` will use the protocol MultiResolver
+      return (resolver as any)[func].bind(resolver, ...args, this.prepareNext(func, next, ...args), this.protocols);
+    } else {
+      return (this.protocols as any)[func].bind(this.protocols, ...args);
+    }
+  }
+
+  async supportsDescriptor(descriptor: Descriptor, opts: MinimalResolveOptions): Promise<boolean> {
+    return await this.prepareNext('supportsDescriptor', this.modules[0], descriptor, opts)();
+  }
+  async supportsLocator(locator: Locator, opts: MinimalResolveOptions): Promise<boolean> {
```

Comment:
Why do we need to turn synchronous methods into asynchronous? Does it have a perf impact? I'd tend to prefer to have an asynchronous `init` step and keep the `supports` synchronous, if it's only that.

Reply:
the idea was that a plugin might need to query a remote repo to see if it supports the descriptor. But in this context "supports" doesn't mean that it can support the specified "transport" (ie. npm, git, tgz, etc.) but that it actually has the package and can respond. Essentially this is what was meant to allow arbitrarily injecting various repos and letting the first one that can respond respond. So (for example) in my plugin I would always check the cache directory first and if it is there respond that it is supported.

Reply:
it would also allow redundancy in the form of fallback repos (that is the idea at least)